File: mobile/android/app/src/main/java/com/tolkflip/network/ApiService.java
package com.tolkflip.network;

import com.tolkflip.models.User;
import com.tolkflip.models.Message;

import java.util.List;
import java.util.Map;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.*;

public interface ApiService {

    // Authentication endpoints
    @POST("/api/auth/request-code")
    Call<Map<String, Object>> requestVerificationCode(@Body Map<String, String> body);

    @POST("/api/auth/verify")
    Call<Map<String, Object>> verifyCode(@Body Map<String, String> body);

    @POST("/api/auth/refresh")
    Call<Map<String, Object>> refreshToken(@Body Map<String, String> body);

    @POST("/api/auth/logout")
    Call<Map<String, Object>> logout(@Body Map<String, String> body);

    // User endpoints
    @GET("/api/users/profile/{userId}")
    Call<Map<String, Object>> getUserProfile(@Path("userId") String userId);

    @PUT("/api/users/profile")
    Call<Map<String, Object>> updateUserProfile(@Body Map<String, Object> body);

    @GET("/api/users/search")
    Call<Map<String, Object>> searchUser(@Query("phone") String phoneNumber);

    @GET("/api/users/contacts")
    Call<Map<String, Object>> getContacts();

    @POST("/api/users/device-token")
    Call<Map<String, Object>> updateDeviceToken(@Body Map<String, String> body);

    // Chat endpoints
    @GET("/api/chat/threads")
    Call<Map<String, Object>> getChatThreads();

    @GET("/api/chat/threads/{threadId}/messages")
    Call<Map<String, Object>> getMessages(
        @Path("threadId") String threadId,
        @Query("limit") int limit,
        @Query("pageState") String pageState
    );

    @POST("/api/chat/threads/{threadId}/settings")
    Call<Map<String, Object>> updateThreadSettings(
        @Path("threadId") String threadId,
        @Body Map<String, Object> settings
    );

    // Translation endpoints
    @POST("/api/translate")
    Call<Map<String, Object>> translateText(@Body Map<String, String> body);

    @POST("/api/translate/batch")
    Call<Map<String, Object>> translateBatch(@Body Map<String, Object> body);

    @GET("/api/translate/languages")
    Call<Map<String, Object>> getSupportedLanguages();

    @POST("/api/translate/detect-language")
    Call<Map<String, Object>> detectLanguage(@Body Map<String, String> body);

    @POST("/api/translate/emotion")
    Call<Map<String, Object>> analyzeEmotion(@Body Map<String, String> body);

    // Transcription endpoints
    @Multipart
    @POST("/api/transcribe")
    Call<Map<String, Object>> transcribeAudio(
        @Part MultipartBody.Part audio,
        @Part("target_language") RequestBody targetLanguage,
        @Part("translate_to") RequestBody translateTo
    );

    // Media endpoints
    @Multipart
    @POST("/api/media/upload")
    Call<Map<String, Object>> uploadMedia(
        @Part MultipartBody.Part file,
        @Part("thread_id") RequestBody threadId,
        @Part("message_id") RequestBody messageId
    );

    @GET("/api/media/media/{mediaId}")
    Call<Map<String, Object>> getMediaUrl(
        @Path("mediaId") String mediaId,
        @Query("thumbnail") boolean thumbnail
    );

    @DELETE("/api/media/media/{mediaId}")
    Call<Map<String, Object>> deleteMedia(@Path("mediaId") String mediaId);

    // Presence endpoints
    @POST("/api/presence/online")
    Call<Map<String, Object>> setUserOnline(@Body Map<String, String> body);

    @POST("/api/presence/offline")
    Call<Map<String, Object>> setUserOffline();

    @GET("/api/presence/user/{userId}")
    Call<Map<String, Object>> getUserPresence(@Path("userId") String userId);

    @POST("/api/presence/users/batch")
    Call<Map<String, Object>> getBatchPresence(@Body Map<String, Object> body);

    @POST("/api/presence/typing")
    Call<Map<String, Object>> updateTypingStatus(@Body Map<String, Object> body);
}

File: mobile/ios/Tolkflip/Services/APIService.swift
import Foundation
import Alamofire
import SwiftyJSON

class APIService {
    static let shared = APIService()

    private let baseURL: String
    private var session: Session

    private init() {
        #if DEBUG
        baseURL = "http://localhost:3000/api"
        #else
        baseURL = "https://api.tolkflip.com/api"
        #endif

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 30

        session = Session(configuration: configuration)
    }

    private func headers() -> HTTPHeaders {
        var headers: HTTPHeaders = [
            "Content-Type": "application/json",
            "Accept": "application/json"
        ]

        if let token = TokenManager.shared.accessToken {
            headers["Authorization"] = "Bearer \(token)"
        }

        return headers
    }

    // MARK: - Authentication

    func requestVerificationCode(phoneNumber: String, completion: @escaping (Result<JSON, Error>) -> Void) {
        let parameters: [String: Any] = ["phone_number": phoneNumber]

        session.request("\(baseURL)/auth/request-code",
                       method: .post,
                       parameters: parameters,
                       encoding: JSONEncoding.default,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    func verifyCode(phoneNumber: String, code: String, displayName: String?,
                   primaryLanguage: String?, completion: @escaping (Result<JSON, Error>) -> Void) {
        var parameters: [String: Any] = [
            "phone_number": phoneNumber,
            "code": code
        ]

        if let name = displayName {
            parameters["display_name"] = name
        }

        if let lang = primaryLanguage {
            parameters["primary_language"] = lang
        }

        session.request("\(baseURL)/auth/verify",
                       method: .post,
                       parameters: parameters,
                       encoding: JSONEncoding.default,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    // MARK: - User Management

    func getUserProfile(userId: String, completion: @escaping (Result<JSON, Error>) -> Void) {
        session.request("\(baseURL)/users/profile/\(userId)",
                       method: .get,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    func updateUserProfile(displayName: String?, avatarUrl: String?,
                          primaryLanguage: String?, additionalLanguages: [String]?,
                          completion: @escaping (Result<JSON, Error>) -> Void) {
        var parameters: [String: Any] = [:]

        if let name = displayName {
            parameters["display_name"] = name
        }

        if let avatar = avatarUrl {
            parameters["avatar_url"] = avatar
        }

        if let lang = primaryLanguage {
            parameters["primary_language"] = lang
        }

        if let langs = additionalLanguages {
            parameters["additional_languages"] = langs
        }

        session.request("\(baseURL)/users/profile",
                       method: .put,
                       parameters: parameters,
                       encoding: JSONEncoding.default,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    // MARK: - Chat

    func getChatThreads(completion: @escaping (Result<JSON, Error>) -> Void) {
        session.request("\(baseURL)/chat/threads",
                       method: .get,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    func getMessages(threadId: String, limit: Int = 50, pageState: String? = nil,
                    completion: @escaping (Result<JSON, Error>) -> Void) {
        var parameters: [String: Any] = ["limit": limit]

        if let state = pageState {
            parameters["pageState"] = state
        }

        session.request("\(baseURL)/chat/threads/\(threadId)/messages",
                       method: .get,
                       parameters: parameters,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    func updateThreadSettings(threadId: String, preferredLanguage: String?,
                             showOriginal: Bool?, enableEmotionDetection: Bool?,
                             completion: @escaping (Result<JSON, Error>) -> Void) {
        var parameters: [String: Any] = [:]

        if let lang = preferredLanguage {
            parameters["preferredLanguage"] = lang
        }

        if let show = showOriginal {
            parameters["showOriginal"] = show
        }

        if let emotion = enableEmotionDetection {
            parameters["enableEmotionDetection"] = emotion
        }

        session.request("\(baseURL)/chat/threads/\(threadId)/settings",
                       method: .post,
                       parameters: parameters,
                       encoding: JSONEncoding.default,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    // MARK: - Translation

    func translateText(text: String, sourceLanguage: String, targetLanguage: String,
                      detectEmotion: Bool = true, completion: @escaping (Result<JSON, Error>) -> Void) {
        let parameters: [String: Any] = [
            "text": text,
            "source_language": sourceLanguage,
            "target_language": targetLanguage,
            "detect_emotion": detectEmotion
        ]

        session.request("\(baseURL)/translate",
                       method: .post,
                       parameters: parameters,
                       encoding: JSONEncoding.default,
                       headers: headers())
            .validate()
            .responseData { response in
                self.handleResponse(response, completion: completion)
            }
    }

    // MARK: - Media Upload

    func uploadMedia(fileData: Data, mimeType: String, threadId: String, messageId: String,
                    completion: @escaping (Result<JSON, Error>) -> Void) {
        let url = "\(baseURL)/media/upload"

        session.upload(multipartFormData: { multipartFormData in
            multipartFormData.append(fileData, withName: "file",
                                   fileName: "media.\(self.fileExtension(for: mimeType))",
                                   mimeType: mimeType)

            if let threadData = threadId.data(using: .utf8) {
                multipartFormData.append(threadData, withName: "thread_id")
            }

            if let messageData = messageId.data(using: .utf8) {
                multipartFormData.append(messageData, withName: "message_id")
            }
        }, to: url, headers: headers())
        .validate()
        .responseData { response in
            self.handleResponse(response, completion: completion)
        }
    }

    // MARK: - Helper Methods

    private func handleResponse(_ response: AFDataResponse<Data>,
                              completion: @escaping (Result<JSON, Error>) -> Void) {
        switch response.result {
        case .success(let data):
            do {
                let json = try JSON(data: data)
                completion(.success(json))
            } catch {
                completion(.failure(error))
            }
        case .failure(let error):
            completion(.failure(error))
        }
    }

    private func fileExtension(for mimeType: String) -> String {
        switch mimeType {
        case "image/jpeg": return "jpg"
        case "image/png": return "png"
        case "image/gif": return "gif"
        case "video/mp4": return "mp4"
        case "audio/mpeg": return "mp3"
        case "audio/wav": return "wav"
        default: return "bin"
        }
    }
}
